sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala

```diff
@@ -331,6 +332,268 @@ case class RoundCeil(child: Expression, scale: Expression)
+  """,
+  arguments = """
+    Arguments:
+      * number - number to be truncated
+      * position - number of decimal places up to which the given number is to be truncated
+    Arguments: To truncate date value:
+      * date - date value or valid date string
+      * fmt - the format representing the unit to be truncated to
+          - "YEAR", "YYYY", "YY" - truncate to the first date of the year that the `date` falls in
+          - "QUARTER" - truncate to the first date of the quarter that the `date` falls in
+          - "MONTH", "MM", "MON" - truncate to the first date of the month that the `date` falls in
+          - "WEEK" - truncate to the Monday of the week that the `date` falls in
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_('2019-08-04', 'week');
+       2019-07-29
+      > SELECT _FUNC_('2019-08-04', 'quarter');
+       2019-07-01
+      > SELECT _FUNC_('2009-02-12', 'MM');
+       2009-02-01
+      > SELECT _FUNC_('2015-10-27', 'YEAR');
+       2015-01-01
+      > SELECT _FUNC_(-10.11, 0);
+       -10
+      > SELECT _FUNC_(10.11, -1);
+       10
+      > SELECT _FUNC_(100.61, 0);
+       100
+      > SELECT _FUNC_(-19087.1560, -3);
+       -19000
+      > SELECT _FUNC_(10876.5489, -1);
+       10870
+      > SELECT _FUNC_(-7767.1160, 2);
+       -7767.11
+      > SELECT _FUNC_(17646.6019, 3);
+       17646.601
+  """,
+  since = "3.4.0",
+  group = "math_funcs")
+// scalastyle:on line.size.limit
+object TruncExpressionBuilder extends ExpressionBuilder {
+  override def build(funcName: String, expressions: Seq[Expression]): Expression = {
+    val numArgs = expressions.length
+    if (numArgs >= 1) {
+      expressions(0).dataType match {
+        case ByteType |  ShortType | IntegerType | LongType | FloatType | DoubleType
+             | DecimalType.Fixed(_, _) => buildTruncNumber(funcName, expressions)
+        case _ => buildTruncDate(funcName, expressions)
+      }
+    } else {
+      throw QueryCompilationErrors.invalidFunctionArgumentNumberError(Seq(2), funcName, numArgs)
+    }
+  }
+
+  private def buildTruncDate(funcName: String, expressions: Seq[Expression]) = {
+    val numArgs = expressions.length
+    if (numArgs == 2) {
+      TruncDate(expressions(0), expressions(1))
+    } else {
+      throw QueryCompilationErrors.invalidFunctionArgumentNumberError(Seq(2), funcName, numArgs)
+    }
+  }
+
+  private def buildTruncNumber(funcName: String, expressions: Seq[Expression]) = {
+    val numArgs = expressions.length
+    if (numArgs >= 1) {
+      val position = if (numArgs == 2) {
+        val positionExpr = expressions(1)
+        if (!(positionExpr.foldable && positionExpr.dataType == IntegerType)) {
+          throw QueryCompilationErrors.requireLiteralParameter(funcName, "position", "int")
+        }
+        val scale_value = positionExpr.eval()
+        if (scale_value == null) {
+          throw QueryCompilationErrors.requireLiteralParameter(funcName, "position", "int")
+        } else {
```

Comment:
no need for an `else` here since throwing the exception will leave the function

Reply:
Done